Updating the strides to be F-continuous for nonzero_static.

mansiag05 opened the pull request:
The issue is different implementation of nonzero_static, producing different stride patterns, causing failures in torch.compile when the fake tensor meta kernel didn't match the real implementation. The nonzero_static, Created C-contiguous tensor (size, ndim) with strides (ndim, 1) and regular nonzero, always uses Fortran-contiguous (1, total_nonzero) stride.

Fixes #152634 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

mansiag05 committed:
Updating the strides to be F-continuous for nonzero_static.

The issue is different implementation of nonzero_static, producing different stride patterns, causing failures in torch.compile when the fake tensor meta kernel didn't match the real implementation.
The nonzero_static, Created C-contiguous tensor (size, ndim) with strides (ndim, 1) and regular nonzero, always uses Fortran-contiguous (1, total_nonzero) stride.

eellison reviewed (changes requested):
please adjust only the fake impl.

mansiag05 committed:
Making nonzero static fake implementation device-aware to match eager behavior. Updated torch/_meta_registrations.py to check fake_device for CPU vs CUDA.
Added device-specific stride tests in test_unary_ufuncs.py.

mansiag05 committed:
Updating the condition for device check.

isuruf reviewed (dismissed):
The test checks eager behaviour which is not what's changing in this PR. Can you add a check for compile? A test case in `test/inductor/test_torchinductor.py` would be good to check that strides match in eager and inductor.

mansiag05 committed:
Test: Add compile check for nonzero_static stride preservation.

Adds test_nonzero_static_stride_compile to verify strides match
between eager and compiled execution for nonzero_static.

mansiag05 committed:
[test] Update stride test for nonzero_static

Update test to verify that nonzero_static produces consistent strides
between eager and fake tensor modes. This ensures the meta kernel
implementation matches the real implementation.

mansiag05 committed:
[test] Skip test_nonzero_static_stride on MPS (operator not implemented)

mansiag05 commented:
@pytorchbot merge

isuruf commented:
@pytorchbot merge

isuruf commented:
@pytorchbot merge -r

isuruf commented:
@pytorchbot merge -f "Unrelated failures"
